packages/flutter_tools/lib/src/web/file_generators/js/flutter_service_worker.js

```diff
@@ -2,6 +2,7 @@
 const MANIFEST = 'flutter-app-manifest';
 const TEMP = 'flutter-temp-cache';
 const CACHE_NAME = 'flutter-app-cache';
+const BASE_HREF = (self.location.pathname.split('/').slice(0, -1).join('/') + '/').substring(1);
```

Comment:
I'm not sure if this works in all scenarios. Let's read it from the `<base>` element like we do here:
https://github.com/flutter/flutter/blob/59ea58b293dfd7f8823e1f8b84c745ad6e81ef5f/packages/flutter_tools/lib/src/web/file_generators/js/flutter.js#L13-L25

Reply:
unfortunately service worker cannot access `document`.

The expression

```
BASE_HREF = (self.location.pathname.split('/').slice(0, -1).join('/') + '/').substring(1);
```

was tested under the following scenarios (producing the following output):

| input                               | output     |
|-------------------------------------|------------|
| (empty)                             | (empty)    |
| /                                   | (empty)    |
| /service_worker.js?v=12345          | (empty)    |
| /something/service_worker.js?v12345 | something/ |
| /a/b/c/service_worker.js?v12345     | a/b/c/     |

always returning base-href without heading slash.

Reply:
My concern with this approach is that you are assuming the service worker is always downloaded from the "root" of the base-href.

Counter example: say the base-href is `/a/b/c/` and we download the service worker from `service/worker.js`. The service worker will be downloaded from `/a/b/c/service/worker.js` and your code will assume that the base-href is `/a/b/c/service/` which is wrong.


> unfortunately service worker cannot access `document`.

This is a bummer 😞 I guess we have no other option then.

Reply:
@mdebbar since this is happening at file generation time, could we read the BASE from a generated index.html file and pass it as a parameter to the service worker generator?

Reply:
> Counter example: say the base-href is /a/b/c/ and we download the service worker from service/worker.js. The service worker will be downloaded from /a/b/c/service/worker.js and your code will assume that the base-href is /a/b/c/service/ which is wrong.

You are right, although in case of such refactoring the e2e test I added would fail which is a good thing 🤔

Reply:
Inspired by @ditman's suggestion, here's an idea: https://github.com/flutter/flutter/commit/89fc35c6de6bfdb79f8bf74c339f767b55985ab7

Would that work?

Reply:
Hmm actually on second thought, that won't work if the user hardcodes the base href in their `index.html` file.